### PR TITLE
[DCP] Reuse partial output in natural-log LSE merge

### PR DESCRIPTION
**Pull request by tanth47**
## Motivation

`cp_lse_ag_out_rs_mha` merges rank-local attention outputs and LSEs per query head. The previous expression allocated a second full FP32 `[tokens, heads, output_dim]` tensor even though the local partial output is not used after the merge.

This merge is not specific to the model attention architecture: `output_dim` can be MHA `v_head_dim` or absorbed-MLA `kv_lora_rank`. In the Kimi-K3 TP8/DCP8 cached-extend reproduction, the absorbed MLA partial had shape `[tokens, 96, 512]`, with `512 = kv_lora_rank`. The extra allocation was 3.000 GiB at 16,384 tokens and caused an OOM on MI350X.

Related to #29736.

## Modifications

- Sanitize `cp_attn_out` in place with `nan_to_num_`.
- Apply the LSE correction in place with `mul_`.
- Reuse the same buffer as the all-reduce input.
- Keep the LSE math, collective, head slicing, and world-size-one path unchanged.

## Accuracy Tests

Real-weight Kimi-K3 DCP8 cap8 versus cap16 validation, using identical datasets and temperature 0:

| Benchmark | cap8 | cap16 | Delta |
| --- | ---: | ---: | ---: |
| GSM8K, 100 examples, 5-shot | 1.000 | 1.000 | 0.000 |
| MMLU, deterministic 100-example sample, zero-shot | 0.910 | 0.910 | 0.000 |

No request failed, and the post-run scan found no OOM, NaN, traceback, runtime error, exception, or fatal event.

The reviewer-triggered `test/registered/dcp/*` suite also passed on Ubuntu, H200, and B200.

## Speed Tests and Profiling

This is a peak-memory fix, not a throughput claim.

- Baseline 16K cached extend: OOM while requesting the 3.00 GiB temporary.
- In-place 16K cached extend: completed with HTTP 200.
- With a 16K scheduler chunk cap, a 32K cached suffix completed as two 16K chunks.

A single 32K prefill is not claimed safe: after removing this temporary, the next observed limit is the DCP Q all-gather.

## Checklist

- [x] Ran pre-commit on the core change.
- [x] Ran the registered DCP suite.
- [x] Ran real-weight DCP8 stability and accuracy A/B tests.

---
### CI States

Latest PR Test (Base): :x: [Run #31403297284](https://github.com/sgl-project/sglang/actions/runs/31403297284)
Latest PR Test (Extra): :x: [Run #31403295225](https://github.com/sgl-project/sglang/actions/runs/31403295225)

### CODE REVIEW

File: test/registered/dcp/test_dcp_mha_lse_merge_unit.py
Comment (kpham-sgl) [inline test/registered/dcp/test_dcp_mha_lse_merge_unit.py:1]: We can remove this test for now, I will come up with a way to test both mla and mha lse merge later

File: python/sglang/srt/layers/dcp/comm.py [hunk 1]
```diff
@@ -85,7 +85,12 @@ def cp_lse_ag_out_rs_mha(
     cp_group: GroupCoordinator,
     return_lse: bool = False,
 ):
-    """Merge DCP partial attention outputs using natural-log LSE (PR #25090)."""
+    """Merge DCP partial attention outputs using natural-log LSE (PR #25090).
+
+    When world_size > 1, this consumes cp_attn_out and reuses its storage for
+    the LSE-corrected local contribution. Callers must not reuse the rank-local
+    partial after this function returns.
+    """
```
Comment (kpham-sgl) [inline python/sglang/srt/layers/dcp/comm.py]: Can just remove this docstring altogether I think

File: python/sglang/srt/layers/dcp/comm.py [hunk 2]
```diff
@@ -95,7 +100,10 @@ def cp_lse_ag_out_rs_mha(
     scale = torch.exp(cp_attn_lse - global_lse).unsqueeze(-1)
     scale = torch.nan_to_num(scale, nan=0.0, posinf=0.0, neginf=0.0)
 
-    out = torch.nan_to_num(cp_attn_out, nan=0.0, posinf=0.0, neginf=0.0) * scale
+    # The rank-local partial is dead after correction. Reuse it instead of
+    # allocating another full [tokens, heads, value_dim] output tensor.
```
Comment (kpham-sgl) [inline python/sglang/srt/layers/dcp/comm.py]: remove AI gen comment

**Review by kpham-sgl — COMMENTED** [review]
FYI, some attention backends use base 2, some use base e for lse. Unsure if we need something similar to `is_mla_dcp_lse_base_on_e` for mha? Out of scope of this PR but just pondering

**Comment by kpham-sgl** [comment]
@tanth47 actually I am a bit confused, why does `cp_lse_ag_out_rs_mha` affect Kimi-K3 🤔

**Comment by tanth47** [comment]
> @tanth47 actually I am a bit confused, why does cp_lse_ag_out_rs_mha affect Kimi-K3 🤔

@kpham-sgl `cp_lse_ag_out_rs_mha` is not actually MHA-specific: it merges rank-local outputs and LSEs per query head, regardless of whether the output dimension is `v_head_dim` or MLA’s `kv_lora_rank`.

Kimi-K3’s cached-extend path uses absorbed MLA, but Triton DCP still calls this helper for the cross-rank merge. The failing tensor was `[16384, 96, 512]`, with `512 = kv_lora_rank`, which confirms the MLA path.

File: python/sglang/srt/layers/dcp/comm.py (hunk 2 above)
Comment (tanth47) [reply python/sglang/srt/layers/dcp/comm.py]: Removed.

File: test/registered/dcp/test_dcp_mha_lse_merge_unit.py
Comment (tanth47) [reply test/registered/dcp/test_dcp_mha_lse_merge_unit.py:1]: Removed as requested. A shared test covering both merge variants sounds better.

File: python/sglang/srt/layers/dcp/comm.py (hunk 1 above)
Comment (tanth47) [reply python/sglang/srt/layers/dcp/comm.py]: Removed.
